Add `sum`, `slice`, `slicesum` to `Hist1d`

dachengx opened the pull request:
1. The `sum` of `Hist1d` and `Histdd` is similar. But `sum` of `Hist1d` sums the histogram and assigns the `bin_edges` as the first and last edges of the old `bin_edges`.
2. `slice` of `Hist1d` takes the histogram between the `start` and `stop` (they should be within the `bin_edges`).
3. `slicesum` runs `slice` first, then `sum`.

dachengx committed:
Add `sum`, `slice`, `slicesum` to `Hist1d`

JelleAalbers reviewed (commented):
Hi Dacheng, thanks for the PR! Indeed, would be good to have these features also in Hist1d.
  * Could you add a description / demonstration?
  * There is some duplication between the newly proposed code and Histdd code. Maybe you can factor out some code to the MultiHistBase base class, or if not, some common function that is called? 
  * The slice and slicesum docstrings say they return a d / d-1 dimensional histogram, that doesn't seem true. I guess slicesum should return a number and slice returns a 1d hist.
  * There is potential interaction with Zihao's PR here https://github.com/JelleAalbers/multihist/pull/17

dachengx committed:
Common function both Hist1d and Histdd can use

dachengx commented:
> Hi Dacheng, thanks for the PR! Indeed, would be good to have these features also in Hist1d.
> 
> * Could you add a description / demonstration?
> * There is some duplication between the newly proposed code and Histdd code. Maybe you can factor out some code to the MultiHistBase base class, or if not, some common function that is called?
> * The slice and slicesum docstrings say they return a d / d-1 dimensional histogram, that doesn't seem true. I guess slicesum should return a number and slice returns a 1d hist.
> * There is potential interaction with Zihao's PR here #17

Hey @JelleAalbers . Thank you for the comments.
 * I added the  description of this PR
 * I copied the common code to be a common function for both `Hist1d` and `Histdd`
 * I fixed this mistake
 * Maybe we can get back to that after this PR is merged

JelleAalbers committed:
Clarify docstrings

JelleAalbers reviewed (approved):
Thanks Dacheng! This will help compatibility between Hist1d and Histdd. Since you need this for a project so I'll make a patch release with this next.
